ospackage/recipes/artifacts.go

```diff
@@ -0,0 +1,186 @@
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"runtime"
+	"strconv"
+	"strings"
+
+	"cloud.google.com/go/storage"
+	osconfigpb "github.com/GoogleCloudPlatform/osconfig/_internal/gapi-cloud-osconfig-go/google.golang.org/genproto/googleapis/cloud/osconfig/v1alpha2"
+)
+
+var (
+	testStorageClient *storage.Client
+	testHTTPClient    *http.Client
+)
+
+func newStorageClient(ctx context.Context) (*storage.Client, error) {
+	if testStorageClient != nil {
+		return testStorageClient, nil
+	}
+	return storage.NewClient(ctx)
+}
+
+func newHTTPClient() *http.Client {
+	if testHTTPClient != nil {
+		return testHTTPClient
+	}
+	return &http.Client{}
+}
+
+// FetchArtifacts takes in a slice of artifacs and dowloads them into the specified directory,
+// Returns a map of artifact names to their new locations on the local disk.
+func FetchArtifacts(ctx context.Context, artifacts []*osconfigpb.SoftwareRecipe_Artifact, directory string) (map[string]string, error) {
+	localNames := make(map[string]string)
+
+	for _, a := range artifacts {
+		path, err := fetchArtifact(ctx, a, directory)
+		if err != nil {
+			return nil, err
+		}
+		localNames[a.Id] = path
+	}
+
+	return localNames, nil
+}
+
+func fetchArtifact(ctx context.Context, artifact *osconfigpb.SoftwareRecipe_Artifact, directory string) (string, error) {
+	localPath := path.Join(directory, artifact.Id)
+	uri, err := url.Parse(artifact.Uri)
+	if err != nil {
+		return "", fmt.Errorf("Could not parse url %q for artifact %q", artifact.Uri, artifact.Id)
+	}
+
+	var reader io.ReadCloser
+	switch strings.ToLower(uri.Scheme) {
+	case "gcs":
+		reader, err = fetchWithGCS(ctx, uri.Host, uri.Path, uri.Fragment)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q from GCS: %v", artifact.Id, err)
+		}
+	case "https", "http":
+		reader, err = fetchWithHTTP(ctx, artifact.Uri)
+		defer reader.Close()
+		if err != nil {
+			return "", fmt.Errorf("error fetching artifact %q with http or https: %v", artifact.Id, err)
+		}
+	default:
+		return "", fmt.Errorf("artifact %q has unsupported protocol %s", artifact.Id, uri.Scheme)
+	}
+
+	err = downloadStream(reader, artifact.Checksum, localPath)
+	if err != nil {
+		return "", err
+	}
+	return localPath, nil
+}
+
+func fetchWithGCS(ctx context.Context, bucket, path, generation string) (io.ReadCloser, error) {
+	client, err := newStorageClient(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create storage client: %v", err)
+	}
+	defer client.Close()
+
+	oh := client.Bucket(bucket).Object(path)
+	if generation != "" {
+		generationNumber, err := strconv.ParseInt(generation, 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("couldn't parse gcs generation number %q", generation)
+		}
+		oh = oh.Generation(generationNumber)
+	}
+
+	r, err := oh.NewReader(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return r, nil
+}
+
+func fetchWithHTTP(ctx context.Context, uri string) (io.ReadCloser, error) {
```

Comment:
don't return interfaces

Reply:
I changed it to reference the actual structs in both of the fetchWith methods

Reply:
yes, http.Response.Body is defined in the struct as being anything which satisfies io.ReadCloser, so that it can internally switch between supported handlers. we don't need to expose that information from http package internals, though, and we should not return interfaces.

Reply:
What's the recommendation for the return value then, would *http.Response make sense?

Reply:
I think that's OK